### PR TITLE
Now,solve((2*sqrt(x)/(x + 1)**2 - 1/(sqrt(x)*(x + 1)) - 1/(4*x**(3/2)))/(x + 1), x) don't raised IndexError

### LINKED ISSUE #18226: solve((2*sqrt(x)/(x + 1)**2 - 1/(sqrt(x)*(x + 1)) - 1/(4*x**(3/2)))/(x + 1), x) has raised IndexError
```python
Python console for SymPy 1.5 (Python 2.7.12)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> eq = Eq((2*sqrt(x)/(x + 1)**2 - 1/(sqrt(x)*(x + 1)) - 1/(4*x**(3/2)))/(x + 1), 0)
>>> solve(eq, x)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/solvers/solvers.py", line 1174, in solve
    solution = _solve(f[0], *symbols, **flags)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/solvers/solvers.py", line 1766, in _solve
    checksol(f_num, {symbol: r}, **flags) is not False]
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/solvers/solvers.py", line 285, in checksol
    val = f.subs(sol)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/basic.py", line 1002, in subs
    rv = rv._subs(old, new, **kwargs)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/compatibility.py", line 897, in wrapper
    result = user_function(*args, **kwds)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/basic.py", line 1116, in _subs
    rv = fallback(self, old, new)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/basic.py", line 1088, in fallback
    arg = arg._subs(old, new, **hints)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/compatibility.py", line 897, in wrapper
    result = user_function(*args, **kwds)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/basic.py", line 1116, in _subs
    rv = fallback(self, old, new)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/basic.py", line 1088, in fallback
    arg = arg._subs(old, new, **hints)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/compatibility.py", line 897, in wrapper
    result = user_function(*args, **kwds)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/basic.py", line 1114, in _subs
    rv = self._eval_subs(old, new)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/power.py", line 778, in _eval_subs
    return new**self.exp._subs(old, new)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/expr.py", line 210, in __pow__
    return self._pow(other)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/decorators.py", line 91, in __sympifyit_wrapper
    return func(a, b)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/decorators.py", line 129, in binary_op_wrapper
    return func(self, other)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/expr.py", line 206, in _pow
    return Pow(self, other)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/compatibility.py", line 897, in wrapper
    result = user_function(*args, **kwds)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/power.py", line 301, in __new__
    obj = b._eval_power(e)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/power.py", line 379, in _eval_power
    b = abs(im(b))*S.ImaginaryUnit
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/expr.py", line 171, in __abs__
    return Abs(self)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/compatibility.py", line 897, in wrapper
    result = user_function(*args, **kwds)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/function.py", line 473, in __new__
    result = super(Function, cls).__new__(cls, *args, **options)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/compatibility.py", line 897, in wrapper
    result = user_function(*args, **kwds)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/function.py", line 288, in __new__
    evaluated = cls.eval(*args)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/functions/elementary/complexes.py", line 473, in eval
    arg = signsimp(arg, evaluate=False)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/simplify/simplify.py", line 383, in signsimp
    e = sub_post(sub_pre(expr))
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/simplify/cse_opts.py", line 18, in sub_pre
    e = e.xreplace(reps)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/basic.py", line 1192, in xreplace
    value, _ = self._xreplace(rule)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/basic.py", line 1207, in _xreplace
    a_xr = _xreplace(rule)
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/core/basic.py", line 1214, in _xreplace
    return self.func(*args), True
  File "/base/data/home/apps/s~sympy-live-hrd/20191227t021942.423433632413966915/sympy/sympy/polys/rootoftools.py", line 331, in __new__
    (-degree, degree - 1, index))
IndexError: root index out of [-2, 1] range, got 2

```

### PR DESCRIPTION
**Pull request by mohitacecode**
#### References to other Issues or PRs

Fixes #18226 

#### Brief description of what is fixed or changed
Now,
```
>>> eq = Eq((2*sqrt(x)/(x + 1)**2 - 1/(sqrt(x)*(x + 1)) - 1/(4*x**(3/2)))/(x + 1), 0)
>>> solve(eq, x)
[-0.154700538379252, 2.15470053837925]
```

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/polys/polytools.py [hunk 1]
```diff
@@ -4167,6 +4167,9 @@ def _strict_eq(f, g):
 class PurePoly(Poly):
     """Class for representing pure polynomials. """
 
+    def __init__(self, rep, *gens, **args):
```
Comment (oscarbenjamin) [inline sympy/polys/polytools.py]: Since `PurePoly` is a `Basic` subclass I don't think it should have an `__init__` method.

Comment (jksuom) [reply sympy/polys/polytools.py]: I have understood that a class can have both a `__new__` and an `__init__` method. A subclass could add something in the `__init__` method.

**Comment by mohitacecode** [comment]
why it is failing on many test cases?

**Comment by jksuom** [comment]
> why it is failing on many test cases?

It is not safe to always use the symbol `x`; I forgot that it can be a domain generator. A dummy should probably be used, maybe something like `Dummy('_')`. That can be found out by experimenting. Many tests need not be edited if the generator is changed only when it is not a symbol. Perhaps something like

    if not self.gens[0].is_Symbol:
        self.gens = (Dummy('_'),)

It is probably necessary to run local tests to see what kind of errors will appear. There are some tests where `gens[0]` is an expression that is not a symbol. Those tests have to be edited. It should be possible without losing their validity.

It should also be possible to add this change in a `__new__` method, if `__init__` is not acceptable.

**Comment by jksuom** [comment]
I agree with @oscarbenjamin that this is not the way to go if instances of `PurePoly` are to be proper `Basic` objects, as the preferably should be.

I have been thinking of another approach: No generators are included among the parameters, only the coefficients and the number of (anonymous) generators. The number is currently always one but it is useful to include it in view of future generalizations.
 
The signature could be `PurePoly(coeffs, ngens)` with possibly `**kwargs` added in the future. `coeffs` is an iterable (or nested iterable in case of several variables) and `ngens` is an integer. Sympified parameters will be the `args` of the Basic object.

In practice, pure polynomials are constructed from ordinary polynomials. There could be a `Poly` method `.as_pure()` for that purpose. There could also be a `PurePoly` method `.as_expr(*syms)` that can be used to represent a pure polynomial as an expression in given generator symbols.

**Comment by jksuom** [comment]
@mohitacecode It is not necessary to fix `PurePoly` to solve this issue. There is another (and better) way: Change the generator of `poly` to a dummy before its roots are computed: https://github.com/sympy/sympy/blob/aef572cdebb498b65ba37a1d56d72d7a2a455b61/sympy/solvers/solvers.py#L1678
The original `poly` should be preserved and the call parameter of `roots` should be something like `dummy_poly = poly.subs(poly.gen, Dummy('x'))`.

**Comment by mohitacecode** [comment]
Okay...I am getting error in very few files first I have to resolve them then I will make a pr soon.

**Comment by mohitacecode** [comment]
I think there is some error in master.
`sympy.test("sympy/parsing/tests/test_c_parser.py")` is failing.

```
Traceback (most recent call last):
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3365, in register_function
    func = getattr(lib, item[0])
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 361, in __getattr__
    func = self.__getitem__(name)
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 366, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/tests/test_c_parser.py", line 34, in test_variable
    res1 = SymPyExpression(c_src1, 'c').return_expr()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/sym_expr.py", line 114, in __init__
    self._expr = parse_c(source_code)
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 653, in parse_c
    converter = CCodeConverter('output')
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 85, in __init__
    super(CCodeConverter, self).__init__()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 57, in __init__
    self.index = cin.Index.create()
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 2130, in create
    return Index(conf.lib.clang_createIndex(excludeDecls, 0))
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 141, in __get__
    value = self.wrapped(instance)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3446, in lib
    register_functions(lib, not Config.compatibility_check)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3393, in register_functions
    register (function)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3390, in register
    return register_function(lib, item, ignore_errors)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3371, in register_function
    raise LibclangError(msg)
clang.cindex.LibclangError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens. Please ensure that your python bindings are compatible with your libclang.so version.
_______________________________________________________________________________________________________________________________________________
________________________________________________ sympy/parsing/tests/test_c_parser.py:test_int ________________________________________________
Traceback (most recent call last):
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3365, in register_function
    func = getattr(lib, item[0])
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 361, in __getattr__
    func = self.__getitem__(name)
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 366, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/tests/test_c_parser.py", line 136, in test_int
    res1 = SymPyExpression(c_src1, 'c').return_expr()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/sym_expr.py", line 114, in __init__
    self._expr = parse_c(source_code)
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 653, in parse_c
    converter = CCodeConverter('output')
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 85, in __init__
    super(CCodeConverter, self).__init__()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 57, in __init__
    self.index = cin.Index.create()
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 2130, in create
    return Index(conf.lib.clang_createIndex(excludeDecls, 0))
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 141, in __get__
    value = self.wrapped(instance)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3446, in lib
    register_functions(lib, not Config.compatibility_check)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3393, in register_functions
    register (function)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3390, in register
    return register_function(lib, item, ignore_errors)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3371, in register_function
    raise LibclangError(msg)
clang.cindex.LibclangError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens. Please ensure that your python bindings are compatible with your libclang.so version.
_______________________________________________________________________________________________________________________________________________
_______________________________________________ sympy/parsing/tests/test_c_parser.py:test_float _______________________________________________
Traceback (most recent call last):
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3365, in register_function
    func = getattr(lib, item[0])
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 361, in __getattr__
    func = self.__getitem__(name)
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 366, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/tests/test_c_parser.py", line 207, in test_float
    res1 = SymPyExpression(c_src1, 'c').return_expr()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/sym_expr.py", line 114, in __init__
    self._expr = parse_c(source_code)
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 653, in parse_c
    converter = CCodeConverter('output')
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 85, in __init__
    super(CCodeConverter, self).__init__()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 57, in __init__
    self.index = cin.Index.create()
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 2130, in create
    return Index(conf.lib.clang_createIndex(excludeDecls, 0))
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 141, in __get__
    value = self.wrapped(instance)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3446, in lib
    register_functions(lib, not Config.compatibility_check)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3393, in register_functions
    register (function)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3390, in register
    return register_function(lib, item, ignore_errors)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3371, in register_function
    raise LibclangError(msg)
clang.cindex.LibclangError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens. Please ensure that your python bindings are compatible with your libclang.so version.
_______________________________________________________________________________________________________________________________________________
_____________________________________________ sympy/parsing/tests/test_c_parser.py:test_function ______________________________________________
Traceback (most recent call last):
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3365, in register_function
    func = getattr(lib, item[0])
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 361, in __getattr__
    func = self.__getitem__(name)
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 366, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/tests/test_c_parser.py", line 295, in test_function
    res1 = SymPyExpression(c_src1, 'c').return_expr()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/sym_expr.py", line 114, in __init__
    self._expr = parse_c(source_code)
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 653, in parse_c
    converter = CCodeConverter('output')
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 85, in __init__
    super(CCodeConverter, self).__init__()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 57, in __init__
    self.index = cin.Index.create()
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 2130, in create
    return Index(conf.lib.clang_createIndex(excludeDecls, 0))
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 141, in __get__
    value = self.wrapped(instance)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3446, in lib
    register_functions(lib, not Config.compatibility_check)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3393, in register_functions
    register (function)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3390, in register
    return register_function(lib, item, ignore_errors)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3371, in register_function
    raise LibclangError(msg)
clang.cindex.LibclangError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens. Please ensure that your python bindings are compatible with your libclang.so version.
_______________________________________________________________________________________________________________________________________________
____________________________________________ sympy/parsing/tests/test_c_parser.py:test_parameters _____________________________________________
Traceback (most recent call last):
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3365, in register_function
    func = getattr(lib, item[0])
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 361, in __getattr__
    func = self.__getitem__(name)
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 366, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/tests/test_c_parser.py", line 376, in test_parameters
    res1 = SymPyExpression(c_src1, 'c').return_expr()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/sym_expr.py", line 114, in __init__
    self._expr = parse_c(source_code)
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 653, in parse_c
    converter = CCodeConverter('output')
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 85, in __init__
    super(CCodeConverter, self).__init__()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 57, in __init__
    self.index = cin.Index.create()
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 2130, in create
    return Index(conf.lib.clang_createIndex(excludeDecls, 0))
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 141, in __get__
    value = self.wrapped(instance)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3446, in lib
    register_functions(lib, not Config.compatibility_check)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3393, in register_functions
    register (function)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3390, in register
    return register_function(lib, item, ignore_errors)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3371, in register_function
    raise LibclangError(msg)
clang.cindex.LibclangError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens. Please ensure that your python bindings are compatible with your libclang.so version.
_______________________________________________________________________________________________________________________________________________
___________________________________________ sympy/parsing/tests/test_c_parser.py:test_function_call ___________________________________________
Traceback (most recent call last):
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3365, in register_function
    func = getattr(lib, item[0])
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 361, in __getattr__
    func = self.__getitem__(name)
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 366, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/tests/test_c_parser.py", line 474, in test_function_call
    res1 = SymPyExpression(c_src1, 'c').return_expr()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/sym_expr.py", line 114, in __init__
    self._expr = parse_c(source_code)
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 653, in parse_c
    converter = CCodeConverter('output')
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 85, in __init__
    super(CCodeConverter, self).__init__()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 57, in __init__
    self.index = cin.Index.create()
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 2130, in create
    return Index(conf.lib.clang_createIndex(excludeDecls, 0))
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 141, in __get__
    value = self.wrapped(instance)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3446, in lib
    register_functions(lib, not Config.compatibility_check)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3393, in register_functions
    register (function)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3390, in register
    return register_function(lib, item, ignore_errors)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3371, in register_function
    raise LibclangError(msg)
clang.cindex.LibclangError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens. Please ensure that your python bindings are compatible with your libclang.so version.
_______________________________________________________________________________________________________________________________________________
_______________________________________________ sympy/parsing/tests/test_c_parser.py:test_parse _______________________________________________
Traceback (most recent call last):
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3365, in register_function
    func = getattr(lib, item[0])
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 361, in __getattr__
    func = self.__getitem__(name)
  File "/home/mohit/Downloads/lib/python3.6/ctypes/__init__.py", line 366, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/tests/test_c_parser.py", line 596, in test_parse
    res1 = SymPyExpression('..a.h', 'c').return_expr()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/sym_expr.py", line 114, in __init__
    self._expr = parse_c(source_code)
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 653, in parse_c
    converter = CCodeConverter('output')
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 85, in __init__
    super(CCodeConverter, self).__init__()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/parsing/c/c_parser.py", line 57, in __init__
    self.index = cin.Index.create()
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 2130, in create
    return Index(conf.lib.clang_createIndex(excludeDecls, 0))
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 141, in __get__
    value = self.wrapped(instance)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3446, in lib
    register_functions(lib, not Config.compatibility_check)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3393, in register_functions
    register (function)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3390, in register
    return register_function(lib, item, ignore_errors)
  File "/home/mohit/Downloads/lib/python3.6/site-packages/clang/cindex.py", line 3371, in register_function
    raise LibclangError(msg)
clang.cindex.LibclangError: /home/mohit/Downloads/bin/python: undefined symbol: clang_annotateTokens. Please ensure that your python bindings are compatible with your libclang.so version.

=========================================== tests finished: 0 passed, 7 exceptions, in 0.16 seconds ===========================================
DO *NOT* COMMIT!
False
>>> 
```

**Comment by mohitacecode** [comment]
>PS. Some tests will fail and have to be edited

I am getting many error of this type.What should be edited here.
```
Failed example:
    integrate(exp(x) / (1 + exp(2 * x)))
Expected:
    RootSum(4*_z**2 + 1, Lambda(_i, _i*log(2*_i + exp(x))))
Got:
    RootSum(4*x**2 + 1, Lambda(_i, _i*log(2*_i + exp(x))))

```

File: sympy/polys/polytools.py (hunk 1 above)
Comment (czgdp1807) [reply sympy/polys/polytools.py]: IMO, objects can be both created and initialised in `__new__`, though for clarity we can use `__init__` in subclasses if there isn't going to be any change in `__new__` method. I am not sure which object will be returned by  `__new__` in this case. May it should be checked.

**Comment by czgdp1807** [comment]
Closing in reference to https://github.com/sympy/sympy/issues/18226#issuecomment-624233887

**Pull request closed without merging**